Portenta Vision shield, capture to bitmap SD Card tutorial [PC-926] (#309)
arduino/docs-content: content/hardware/04.pro/shields/portenta-vision-shield/tutorials/camera-to-bitmap-sd-card/content.md

---
title: "Saving Images as Bitmaps Into the SD Card"
difficulty: easy
tags: [Camera, Bitmap, SD Card]
description: This tutorial shows you how to capture a frame from the Portenta Vision Shield Camera module and save the output as a bitmap image.
author: Pablo Marquínez
hardware:
  - hardware/04.pro/boards/portenta-h7
  - hardware/04.pro/shields/portenta-vision-shield
software:
  - ide-v1
  - ide-v2
  - web-editor
  - cli
---

## Overview
This tutorial shows you how to capture a frame from the Portenta Vision Shield Camera module and save the output as a bitmap image. It will allow you to see the output directly on your computer without using any third party tool.

## Goals

- Capturing a frame from the camera.
- Make the bitmap binary file with the correct settings.

> - Save the bitmap on SD Card.

Reviewer: ```suggestion
- Save the bitmap on an SD Card.
```

> - Visualize the image on your computer.

Reviewer: ```suggestion
- Visualize the captured image on your computer.
```

### Required Hardware and Software

- [Portenta H7](https://store.arduino.cc/portenta-h7)
- Portenta Vision Shield ([LoRa](https://store.arduino.cc/portenta-vision-shield-lora) or [Ethernet](https://store.arduino.cc/portenta-vision-shield))
- 1x USB-C cable (either USB-A to USB-C or USB-C to USB-C)
- Micro SD card
- Arduino IDE or Arduino-cli

## Instructions

### 1. The Setup
Connect the Vision Shield to your Portenta H7 as shown in the figure. The top and bottom high density connectors are connected to the corresponding ones on the underside of the H7 board. Plug in the H7 to your computer using the USB C cable.

![Connecting the Vision Shield to Portenta](assets/vs_ard_gs_attach_boards.svg)

#### The Camera

You will be using the **Himax HM-01B0 camera module** which has a resolution of 320 by 240 and the output data its in grayscale with 8 bits per pixel (bpp), this is important to have in mind as the `.bmp` (bitmap) format has some needed configuration depending on the data being used.

> Inside the sketch you can use this librarys to access the camera APIs

Reviewer: ```suggestion
Inside the sketch you can use these libraries to access the camera APIs
```

```cpp
#include "camera.h"
#include "himax.h"
```

#### Bitmap File Format

> The bitmap binary file needs to contain some information in order to tell the computer some required information, for example the resolution of the picture and the bit-depth (bpp)

Reviewer: ```suggestion
The bitmap binary file needs to contain some information in order to tell the computer some required information, for example the resolution of the picture and the bit-depth (bpp). Bit depth refers to the color information stored in the image. The higher the bit depth of an image, the more colors it can store. As the bit depth increases, the file size of the image also increases because more color information has to be stored for each pixel in the image.
```

The following table shows all the headers, the size of its buffer, offsets, the settings that we used and a details column:

| Name                | Size        | Details                                                      |
| ------------------- | ----------- | ------------------------------------------------------------ |
| DIB                 | 14 Bytes    | Bitmap information, setting the size of the file.            |
| File Header         | 40 Bytes    | This header requires the resolution and the bpp               |
| Palette (Color Map) | 1025 Bytes  | This header is mandatory on bitmaps with a bpp ≤ 8, setting the grayscale |
| Image data          | 76800 Bytes | The raw image data, in this case each pixel has 8 bits (1 Byte) and the resolution is 320 x 240, with no compression |

The final size of the file is **77.879 KB**.

### 2. The Sketch

You can find the sketch on the latest version of the [Arduino_Pro_Tutorials](https://github.com/arduino-libraries/Arduino_Pro_Tutorials) at `examples > Vision Shield to SD Card bmp > visionShieldBitmap.ino`

***You can see the full sketch on the [Full sketch section](#full-sketch)***

> First you need to include the needed libraries

Author: Fixed on 510fcb2

```cpp
#include "SDMMCBlockDevice.h"
#include "FATFileSystem.h"

#include "camera.h"
#include "himax.h"
```

> Then we define the block system (`blockDevice` and `fileSystem`) object needed to get access to the SD Card and the the 

Author: Fixed

```cpp
#include "SDMMCBlockDevice.h"
#include "FATFileSystem.h"
SDMMCBlockDevice blockDevice;
mbed::FATFileSystem fileSystem("fs");

#include "camera.h"
#include "himax.h"
HM01B0 himax;
Camera cam(himax);
```

For the bitmap headers binary file we will need some information like the resolution of the image, the bits per pixel and more, so we define our settings as shown:
```cpp
// Settings for our setup
#define RES_H (unsigned int)240
#define RES_W (unsigned int)320
#define IMAGE_MODE CAMERA_GRAYSCALE
#define IMAGE_BPP (unsigned int)8
// Headers info
#define HEADER_FILE_HEADER (unsigned int)14
#define HEADER_DIB_SIZE (unsigned int)40
#define HEADER_FULL_SIZE (HEADER_FILE_HEADER + HEADER_DIB_SIZE)
#define PALETTE_SIZE (2 ^ IMAGE_BPP) * 4 // 4 bytes per color
```

To mount the SD Card the sketch has one function to do so:
```cpp
// Mount File system block
void mountSD()
{
    Serial.println("Mounting SD Card...");

    int error = fileSystem.mount(&blockDevice);
    if (error)
    {
        Serial.println("No SD Card found");
        while (1)
            ;
    }
}
```

> Another function that the sketch contains is the one to generate the `.bmp` file

Author: Fixed

```cpp
#include "SDMMCBlockDevice.h"
#include "FATFileSystem.h"
SDMMCBlockDevice blockDevice;
mbed::FATFileSystem fileSystem("fs");

#include "camera.h"
#include "himax.h"
HM01B0 himax;
Camera cam(himax);
```

For the bitmap headers binary file we will need some information like the resolution of the image, the bits per pixel and more, so we define our settings as shown:
```cpp
// Settings for our setup
#define RES_H (unsigned int)240
#define RES_W (unsigned int)320
#define IMAGE_MODE CAMERA_GRAYSCALE
#define IMAGE_BPP (unsigned int)8
// Headers info
#define HEADER_FILE_HEADER (unsigned int)14
#define HEADER_DIB_SIZE (unsigned int)40
#define HEADER_FULL_SIZE (HEADER_FILE_HEADER + HEADER_DIB_SIZE)
#define PALETTE_SIZE (2 ^ IMAGE_BPP) * 4 // 4 bytes per color
```

To mount the SD Card we will use the following function in the sketch:
```cpp
// Mount File system block
void mountSD()
{
    Serial.println("Mounting SD Card...");

    int error = fileSystem.mount(&blockDevice);
    if (error)
    {
        Serial.println("No SD Card found");
        while (1)
            ;
    }
}
```

Another function that the sketch contains is the one to generate the `.bmp` file called `parseData()`.

> This function will create the needed headers that later we will need to encode our bitmap into a file.

Reviewer: ```suggestion
This function will create the needed headers that later we will need later to encode our bitmap into a file.
```

Author: Get rid of the first "later" 🤔 
```suggestion
This function will create the needed headers that we will need later to encode our bitmap into a file.
```

> ***Take a look to the section [Bitmap File Format](#bitmap-file-format) to better understand the file headers that are created on this function***

Reviewer: ```suggestion
***Take a look at the section [Bitmap File Format](#bitmap-file-format) to better understand the file headers that are created with this function***
```

| DIB                 | 14 Bytes    | Bitmap information, setting the size of the file.            |
| File Header         | 40 Bytes    | This header requires the resolution and the bpp               |
| Palette (Color Map) | 1025 Bytes  | This header is mandatory on bitmaps with a bpp ≤ 8, setting the grayscale |
| Image data          | 76800 Bytes | The raw image data, in this case each pixel has 8 bits (1 Byte) and the resolution is 320 x 240, with no compression |

The final size of the file is **77.879 KB**.

### 2. The Sketch

You can find the sketch on the latest version of the [Arduino_Pro_Tutorials](https://github.com/arduino-libraries/Arduino_Pro_Tutorials) at `examples > Vision Shield to SD Card bmp > visionShieldBitmap.ino`

***You can see the full sketch in the [Full sketch section](#full-sketch)***

First you need to include the needed libraries
```cpp
#include "SDMMCBlockDevice.h"   // Multi Media Card APIs
#include "FATFileSystem.h"      // Mbed API for portable  and embedded systems

#include "camera.h" // Arduino Mbed Core Camera APIs
#include "himax.h"  // Exclusive Camera library for the Portenta Vision Shield
```

> Then we define the following objects with their respective constructor (`blockDevice` and `fileSystem` objects), needed for getting access to the SD Card and the the file system.

Reviewer: ```suggestion
Then we define the following objects with their respective constructor (`blockDevice` and `fileSystem` objects), needed for getting access to the SD Card and the file system.
```
